worker/uniter/runner/context/context.go

```diff
@@ -985,116 +989,86 @@ func (ctx *HookContext) Prepare() error {
 }
 
 // Flush implements the runner.Context interface.
-func (ctx *HookContext) Flush(process string, ctxErr error) (err error) {
-	writeChanges := ctxErr == nil
+func (ctx *HookContext) Flush(process string, ctxErr error) error {
+	// Apply the changes if no error reported while the hook was executing.
+	var flushErr error
+	if ctxErr == nil {
+		flushErr = ctx.doFlush(process)
+	}
 
-	// In the case of Actions, handle any errors using finalizeAction.
 	if ctx.actionData != nil {
-		// If we had an error in err at this point, it's part of the
-		// normal behavior of an Action.  Errors which happen during
-		// the finalize should be handed back to the uniter.  Close
-		// over the existing err, clear it, and only return errors
-		// which occur during the finalize, e.g. API call errors.
-		defer func(ctxErr error) {
-			err = ctx.finalizeAction(ctxErr, err)
-		}(ctxErr)
-		ctxErr = nil
-	} else {
-		// TODO(gsamfira): Just for now, reboot will not be supported in actions.
-		defer ctx.handleReboot(&err)
+		// While an Action is executing, errors may happen as part of
+		// its normal flow. We pass both potential action errors
+		// (ctxErr) and any flush errors to finalizeAction helper
+		// which will figure out if an error needs to be returned back
+		// to the uniter.
+		return ctx.finalizeAction(ctxErr, flushErr)
 	}
 
+	// TODO(gsamfira): Just for now, reboot will not be supported in actions.
+	if ctxErr == nil {
+		ctxErr = flushErr
+	}
+	return ctx.handleReboot(ctxErr)
+}
+
+func (ctx *HookContext) doFlush(process string) error {
+	b := uniter.NewConsolidateHookParamsBuilder(ctx.unit.Tag())
+
 	// When processing config changed hooks we need to ensure that the
 	// relation settings for the unit endpoints are up to date after
 	// potential changes to already bound endpoints.
 	if process == string(hooks.ConfigChanged) {
-		if e := ctx.unit.UpdateNetworkInfo(); e != nil {
-			return errors.Trace(e)
-		}
+		b = b.UpdateNetworkInfo()
```

Comment:
I prefer chaining for writing terser tests (e.g. see how this is applied here https://github.com/juju/juju/pull/11204/files#diff-e63654fb040d13dd920419c8b37aa04fR4630-R4637) but I am happy to change it to multiple calls if others prefer the other suggested style. Also note that gomock uses chaining as well for specifying call expectations.

As far as errors are concerned, I believe the general pattern is that the builder buffers errors internally and yields them when you call the terminal method (Build/Finalize/...).

If the assignment code looks ugly (I _think_ the compiler is smart to optimize the assignment away since it tracks the pointer), I can just drop the `b = ` bit (or change it to `_ = `) with the same effect if that improves readability.

Reply:
Unfortunately the link doesn't actually work cleanly because the PR is too big (and folding, etc hides it from me).
As mentioned above, I did see a test that used chaining, but all that drops is a 'b' on each line and moves the '.' onto the previous line. Which doesn't actually seem 'terser'.
My argument is that you're trading off slightly uglier production code for "terser" test code, that isn't actually terser (IMO).

When *I* read the above line, I have to think "why are we replacing b" but then when I go to the code, we *aren't*, and it adds friction.

I do realize that gomock uses the pattern, and it does lead to terse lines like:
```go
mockBackend.EXPECT().Machine(s.machineTag1.Id()).Return(mockMachine1, nil)
```

However, I don't think we get the same benefit here. The things we're doing don't feel like quick-small-updates that benefit from thinking as all one thing.


I certainly don't think `_ =` would aid clarity. In fact that would look to me like we are explicitly ignoring an error return. I'm not worried about compiler optimization/inefficiency. Its about code readability.

We can certainly bring in a third viewpoint and have someone else weigh in.

Reply:
I don't have a strong preference for either approach. I will change it for now to get rid of chaining since it does in fact make the non-test code a bit more difficult to read